sw/device/boot_rom/bootstrap.c

```diff
@@ -54,35 +56,40 @@ static int erase_flash(void) {
  * hash. Returns 0 if both hashes are identical. */
 static int check_frame_hash(const frame_t *f) {
   static uint32_t hash[8];
-  uint32_t result = 0;
   hw_SHA256_hash((uint8_t *)&f->hdr.frame_num, RAW_BUFFER_SIZE - 32,
                  (uint8_t *)hash);
-  for (int i = 0; i < 8; ++i) {
-    result |= hash[i] ^ f->hdr.hash[i];
-  }
-  return result;
+  LOG_INFO("Hash starts: want 0x%8x, got 0x%8x.", f->hdr.hash[0], hash[0]);
+  return memcmp(hash, f->hdr.hash, sizeof(hash));
 }
 
-/* Processes frames received via spid interface and writes them to flash. */
-static int bootstrap_flash(void) {
-  static frame_t f;
-  static uint8_t ack[32] = {0};
+static int bootstrap_flash(const dif_spi_device_t *spi) {
+  uint8_t ack[32] = {0};
   uint32_t expected_frame_no = 0;
-  for (;;) {
-    if (spid_bytes_available() >= sizeof(f)) {
-      spid_read_nb(&f, sizeof(f));
+  while (true) {
+    size_t bytes_available;
+    CHECKZ(dif_spi_device_rx_pending(spi, &bytes_available),
```

Comment:
FWIW, I also think moving the CHECKZ to a separate line is more readable. I do also wonder if we might as well just rely on the single CHECK macro and do `CHECK(res == 0, "Failed to read from SPI.");`

Reply:
FWIW, I also think moving the CHECKZ to a separate line is more readable. I do also wonder if we might as well just rely on the single CHECK macro and do `CHECK(res == 0, "Failed to read from SPI.");`

Reply:
Silvestrs: the whole reason to have a single CHECK macro definition is so that it is consistently used everywhere for asserting, because C does not have a reasonable assert. If we're going to have CHECK at all, it should be used consistently.

Also, putting the result on another line requires naming the result type, which is the whole thing CHECK is trying to avoid in the first place (because C does not have auto).

Reply:
I see it as an unnecessary complication in this case, we can agree to disagree on this one.